core/desugarFuns.ml

```diff
@@ -67,19 +67,30 @@ class desugar_funs env =
 object (o : 'self_type)
   inherit (TransformSugar.transform env) as super
 
+  method private desugarFunLit argss lin lam location tvs =
+    let inner_mb     = snd (last argss) in
+    let (o, lam, rt) = o#funlit inner_mb lam in
+    let ft = List.fold_right (fun (args, mb) rt -> `Function (args, mb, rt))
+                             argss rt in
+    let f = gensym ~prefix:"_fun_" () in
+    let body, tvs, ft = match tvs with
+      | [] -> Var f, [], ft
+      | _  ->
+         let (tvs, tyargs), ft = Generalise.generalise env.Types.var_env ft in
+         let ft = Instantiate.freshen_quantifiers ft in
```

Comment:
Voodoo coding here. Do I need to call `freshen_quantifiers`? It seems to be called right after `generalise` in other places in the code so I inserted it here as well. But is it always required? (And if yes, why isn't it folded into `generalise`?)

Reply:
My guess would be: necessary if you (or whatever code eventually receives `ft`) are planning on substituting something with free variables into `ft` inside the scope of a binder, but shouldn't hurt either way.  (Alternatively perhaps whatever does the substitution should do this.)

Reply:
I think this is necessary here. `Generalise.generalise` binds type variables without changing their names. 
In your example
```
TAbs Y. { 
  fun freshfun(x:Z) {x}
  TApp freshfun [Y]
}
```
Here, `freshfun` must not have type `forall Y. Y -> Y`, because then we would have two binding occurences for type variable `Y`, which violates an invariant of the IR (unique names). Instead, it must freshen the quantifiers (e.g., to `Z`, as you did in the example).